### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by karlkfi**
## High Level Description

- reorganize for usability
- make headers easier to link to with simpler names
- extract development environment section
- extract unit tests section
- add integration tests section
- add releases section
- clarify pull request status checks

## Checklist for all PR's

  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [X] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [X] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,16 +204,44 @@ storage:
...
+| *pkgpanda*           | DC/OS baseline/host package management system. Tools for building, deploying, upgrading, and bundling packages together which live on the root filesystem of a machine / underneath Mesos.
+| *pytest*             | Misc. tests. Should be moved to live next to the appropriate code
+| *release*            | Release tools for DC/OS. (Building releases, building installers for releases, promoting between channels)
+| *ssh*                | AsyncIO based parallel ssh library used by the installer
+| *test_util*          | various scripts, utilities to help with integration testing
+
+All code in this repository is Python 3
+
+
 # TODO
```
Comment (orsenthil) [inline README.md]: I vote to kill these TODO. we track them in JIRAs.

File: README.md [hunk 2]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+
+## Minimum Requirements
+
+- 1 master node
+- 2 private agent nodes
+- 1 public agent node
+- Task resource allocation is currently insignificantly small
+- DC/OS itself requires at least 2 (virtual) cpu cores on each node
+
+## Instructions
```
Comment (orsenthil) [inline README.md]: +1 to this information on how to run integration tests.

File: README.md (hunk 1 above)
Comment (karlkfi) [inline README.md]: get me another vote and i'll kill them.

File: README.md [hunk 3]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
-| *gen/build_deploy*   | Code to take a build and transform it into a particular platform deployment tool (Bash / command line, AWS, Azure, etc.)
-| *packages*           | Packages which make up DC/OS (Mesos, Marathon, AdminRouter, etc). These packages are built by pkgpanda, and combined into a "bootstrap" tarball for deployment.
-| *pkgpanda*           | DC/OS baseline/host package management system. Tools for building, deploying, upgrading, and bundling packages together which live on the root filesystem of a machine / underneath Mesos.
-| *pytest*             | Misc. tests. Should be moved to live next to the appropriate code
-| *release*            | Release tools for DC/OS. (Building releases, building installers for releases, promoting between channels)
-| *ssh*                | AsyncIO based parallel ssh library used by the installer
-| *test_util*          | various scripts, utilities to help with integration testing
+Release artifacts are managed by Mesosphere on Amazon S3, using a CloudFront cache, under the path <https://downloads.dcos.io/dcos/>
```
Comment (adam-mesos) [inline README.md]: I get AccessDenied when I try to access this URL. Is it worth linking this when we have stable/EA/master links below?

File: README.md [hunk 4]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+
+To find the git SHA of any given release, check the latest commit in the versioned branches on GitHub: <https://github.com/dcos/dcos/branches/>
+
+| Release Type | URL Pattern |
+|--------------|--------------------|
+| Latest Stable| `https://downloads.dcos.io/dcos/stable/dcos_generate_config.sh` |
```
Comment (adam-mesos) [inline README.md]: Why not make this and the EA/master links clickable?

File: README.md [hunk 5]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+# Development Environment
+
+**Linux is required for building and testing DC/OS.**
+
+1. Linux distribution:
+  - Docker doesn't have all the features needed on OS X or Windows
+  - `tar` needs to be GNU tar for the set of flags used
+1. [tox](https://tox.readthedocs.org/en/latest/)
+1. git 1.8.5+
+1. Docker
```
Comment (adam-mesos) [inline README.md]: We now require/recommend Docker 1.11, as of DC/OS 1.9

File: README.md [hunk 6]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
-  1. _Optional_ pxz (speeds up package and bootstrap compression)
-    - ArchLinux: [pxz-git in the AUR](https://aur.archlinux.org/packages/pxz-git). The pxz package corrupts tarballs fairly frequently.
-    - Fedora 23: `sudo dnf install pxz`
-
-## Running local code quality tests
+# Development Environment
+
+**Linux is required for building and testing DC/OS.**
+
+1. Linux distribution:
```
Comment (adam-mesos) [inline README.md]: Minimum kernel version?

File: README.md [hunk 7]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+  - Docker doesn't have all the features needed on OS X or Windows
+  - `tar` needs to be GNU tar for the set of flags used
+1. [tox](https://tox.readthedocs.org/en/latest/)
+1. git 1.8.5+
+1. Docker
+  - [Install Instructions for various distributions](https://docs.docker.com/engine/installation/). Docker needs to be configured so your user can run docker containers. The command `docker run alpine  /bin/echo 'Hello, World!'` when run at a new terminal as your user should just print `"Hello, World!"`. If it says something like "Unable to find image 'alpine:latest' locally" then re-run and the message should go away.
+1. Python 3.5
+  - Arch Linux: `sudo pacman -S python`
```
Comment (adam-mesos) [inline README.md]: Do we support Arch Linux? I suppose maybe as a dev environment, but not for deployments

File: README.md [hunk 8]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+    - [pyenv-installer](https://github.com/yyuu/pyenv-installer)
+    - Python dependencies: `sudo apt-get install make build-essential libssl-dev zlib1g-dev libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev libncursesw5-dev xz-utils liblzma-dev`
+    - Install Python 3.5.2: `pyenv install 3.5.2`
+    - Create DC/OS virtualenv: `pyenv virtualenv 3.5.2 dcos`
+    - Activate environment: `pyenv activate dcos`
+1. Over 10GB of free disk space
+1. _Optional_ pxz (speeds up package and bootstrap compression)
+  - ArchLinux: [pxz-git in the AUR](https://aur.archlinux.org/packages/pxz-git). The pxz package corrupts tarballs fairly frequently.
+  - Fedora 23: `sudo dnf install pxz`
+
```
Comment (adam-mesos) [inline README.md]: Is `pxz` not available on Ubuntu?

File: README.md [hunk 9]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+
+# Unit Tests
+
+Unit tests can be run locally but require the [development environment](#development-environment) specified above.
+
 ```
 tox
 ```
```
Comment (adam-mesos) [inline README.md]: I don't know why this is a code block. Maybe better as a `## Tox` sub-header?

File: README.md [hunk 10]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+
+    ```
+    sudo su -
+    ```
+
+1. Add the test user
+
+    ```
+    source /opt/mesosphere/environment.export
+    python /opt/mesosphere/active/dcos-oauth/bin/dcos_add_user.py albert@bekstil.net
```
Comment (adam-mesos) [inline README.md]: Is albert@bekstil.net the only test user we can use? Is this one of @alberts emails?

File: README.md [hunk 11]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+    ```
+
+    This test user has a known login token with far future expiration. DO NOT USE IN PRODUCTION!
```
Comment (adam-mesos) [inline README.md]: Maybe point to docs on how to remove the user after you've added it to run your post-install integration tests?
I fear that customers might run the integration tests and then forget that they've got an immortal user account (backdoor).

File: README.md [hunk 12]
```diff
@@ -21,74 +23,135 @@ To learn more, see the [DC/OS Overview](https://dcos.io/docs/latest/overview/).
...
+    ```
+
+
+# Build
+
+DC/OS can be built locally but requires the [development environment](#development-environment) specified above.
+
+DC/OS builds are packaged as a self-extracting Docker image wrapped in a bash script called `dcos_generate_config.sh`.
+
+**WARNING**: Building a release from scratch the first time on a modern dev machine (4 cores / 8 hyper threads, SSD, reasonable interent bandwidth) takes about 1 hour.
```
Comment (adam-mesos) [inline README.md]: s/interent/internet/

File: README.md [hunk 13]
```diff
@@ -140,16 +204,44 @@ storage:
...
+- teamcity/code-quality: simply run `tox` in the top-level dir to run all syntax checks as well as pytest (unit-tests). See [tox.ini](https://github.com/dcos/dcos/blob/master/tox.ini) for more details
+- integration-test/*: runs [integration_test.py](https://github.com/dcos/dcos/blob/master/test_util/integration_test.py) in the network of a DC/OS cluster
     - /vagrant-bash: Tests the on-prem bash provider by using [dcos-vagrant](https://github.com/dcos/dcos-vagrant). Invoke this test through [run-all](https://github.com/dcos/dcos/blob/master/test_util/run-all)
     - /deploy-vpc-cli: runs [ccm-deploy-test](https://github.com/dcos/dcos/blob/master/test_util/test_installer_ccm.py) with USE_INSTALLER_API=false. A Virtual Private Cloud of centos nodes is spun up by CCM (Mesosphere's Cloud Cluster Manager) and the installer (dcos_generate_config.sh) is used via the CLI options to deploy DC/OS. Finally, the same integration_test.py is run
```
Comment (adam-mesos) [inline README.md]: s/centos/CentOS/?

File: README.md [hunk 14]
```diff
@@ -140,16 +204,44 @@ storage:
...
+
+# Repo Structure
+
+DC/OS itself is composed of many individual components precisely configured to work together in concert.
+
+This repo contains the release and package building tools necessary to produce installers for various on-premises and cloud platforms.
+
+| Directory | Contents |
+| --------- | -------- |
+| *docker*             | Locally defined docker containers packages are built in
```
Comment (adam-mesos) [inline README.md]: I don't see this directory anywhere. Instead, I see `cloud_images` and `config`

File: README.md [hunk 15]
```diff
@@ -140,16 +204,44 @@ storage:
...
+| *docs*               | Documentation
```
Comment (adam-mesos) [inline README.md]: Technically after `dcos_installer` alphabetically

File: README.md [hunk 16]
```diff
@@ -140,16 +204,44 @@ storage:
...
+| *dcos_installer*     | Backend for Web, SSH, and some bits of the Advanced installer. Code is being cleaned up
+| *gen*                | Python library for rendering yaml config files for various platforms into packages, with utilities to do things like make "late binding" config set by CloudFormation
```
Comment (adam-mesos) [inline README.md]: Skipped the `flake8_dcos_lint` folder

File: README.md [hunk 17]
```diff
@@ -140,16 +204,44 @@ storage:
...
+| *gen/build_deploy*   | Code to take a build and transform it into a particular platform deployment tool (Bash / command line, AWS, Azure, etc.)
```
Comment (adam-mesos) [inline README.md]: Why call out this subdirectory of `gen/`?

File: README.md [hunk 18]
```diff
@@ -140,16 +204,44 @@ storage:
...
+| *packages*           | Packages which make up DC/OS (Mesos, Marathon, AdminRouter, etc). These packages are built by pkgpanda, and combined into a "bootstrap" tarball for deployment.
+| *pkgpanda*           | DC/OS baseline/host package management system. Tools for building, deploying, upgrading, and bundling packages together which live on the root filesystem of a machine / underneath Mesos.
+| *pytest*             | Misc. tests. Should be moved to live next to the appropriate code
```
Comment (adam-mesos) [inline README.md]: This folder seems to have disappeared.

File: README.md (hunk 14 above)
Comment (adam-mesos) [inline README.md]: Ok, seems like it lives in `/pkgpanda/docker/dcos-builder/` now

File: README.md (hunk 1 above)
Comment (adam-mesos) [inline README.md]: +1 to removing the TODOs. I think this PR alone covers:
  - Add getting started on common distros / dependencies
  - Add overview of what is in here, how it works
  - Add general theory of stuff that goes in here.
  - PR (guidelines, testing)

File: README.md [hunk 19]
```diff
@@ -140,16 +204,44 @@ storage:
...
+| *release*            | Release tools for DC/OS. (Building releases, building installers for releases, promoting between channels)
+| *ssh*                | AsyncIO based parallel ssh library used by the installer
+| *test_util*          | various scripts, utilities to help with integration testing
+
+All code in this repository is Python 3
```
Comment (adam-mesos) [inline README.md]: Technically it's only 93.3% Python (3), then 6.3% Shell, and 69 lines of C

File: README.md (hunk 10 above)
Comment (orsenthil) [inline README.md]: Yes.

File: README.md (hunk 19 above)
Comment (orsenthil) [inline README.md]: Contextual. The intention, I assume, was to differentiate Python 2 usage vs Python 3.
I'd be okay with killing this line.

Comment (karlkfi) [inline README.md]: The dev env already lists python 3. so i'll nuke this line

**Comment by karlkfi** [comment]
@adam-mesos I didn't modify the "in this repo" section. I just moved it down to the bottom.
If it needs revamp, you could do it in another PR.

File: README.md (hunk 9 above)
Comment (karlkfi) [inline README.md]: because `tox` is how you run the tests.

File: README.md (hunk 4 above)
Comment (karlkfi) [inline README.md]: I'd rather have people downloading from <dcos.io/releases>. This is more of a reference.

File: README.md (hunk 3 above)
Comment (karlkfi) [inline README.md]: I can change it to an inine code block.

Comment (karlkfi) [inline README.md]: or remove it

File: README.md (hunk 7 above)
Comment (karlkfi) [inline README.md]: I didn't add this. I just moved it.

File: README.md (hunk 8 above)
Comment (karlkfi) [inline README.md]: I didn't add this. I just moved it.

File: README.md (hunk 11 above)
Comment (karlkfi) [inline README.md]: Where in the docs would that be?

File: README.md (hunk 14 above)
Comment (karlkfi) [inline README.md]: I'll remove it

File: README.md (hunk 11 above)
Comment (karlkfi) [inline README.md]: I added a link to https://dcos.io/docs/latest/administration/user-management/

**Comment by karlkfi** [comment]
I also didn't write the dependencies section, just extracted it. If you'd like specific additions, let me know what they should be.

**Comment by orsenthil** [comment]
After approval from @adam-mesos , please make sure to squash these changes into a single commit.

**Comment by karlkfi** [comment]
squashed

**Comment by karlkfi** [comment]
rebased

**Comment by orsenthil** [comment]
Thank you!

**Pull request merged**